// @doist/todoist-api-typescript/dist/cjs/types/sync/resources/completed-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompletedInfoSchema = exports.CompletedTaskMetadataSchema = exports.CompletedSectionMetadataSchema = exports.CompletedProjectMetadataSchema = void 0;
const zod_1 = require("zod");
exports.CompletedProjectMetadataSchema = zod_1.z
    .object({
    projectId: zod_1.z.string(),
    archivedSections: zod_1.z.number().int(),
    completedItems: zod_1.z.number().int(),
})
    .passthrough();
exports.CompletedSectionMetadataSchema = zod_1.z
    .object({
    sectionId: zod_1.z.string(),
    id: zod_1.z.string(),
    completedItems: zod_1.z.number().int(),
})
    .passthrough();
exports.CompletedTaskMetadataSchema = zod_1.z
    .object({
    itemId: zod_1.z.string(),
    id: zod_1.z.string().optional(),
    completedItems: zod_1.z.number().int(),
})
    .passthrough();
exports.CompletedInfoSchema = zod_1.z.union([
    exports.CompletedProjectMetadataSchema,
    exports.CompletedSectionMetadataSchema,
    exports.CompletedTaskMetadataSchema,
]);
