// @doist/todoist-api-typescript/dist/cjs/types/sync/resources/live-notifications.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiveNotificationSchema = void 0;
const zod_1 = require("zod");
/**
 * Live notification resource from the Sync API.
 *
 * Uses a base schema with commonly-present fields plus `.passthrough()` for
 * forward compatibility. The Sync API returns 20+ notification variants;
 * fully typing each is too fragile for a library, so we validate the common
 * shape and preserve all extra fields.
 */
exports.LiveNotificationSchema = zod_1.z
    .object({
    id: zod_1.z.string(),
    createdAt: zod_1.z.string(),
    fromUid: zod_1.z.string(),
    notificationType: zod_1.z.string(),
    isUnread: zod_1.z.boolean(),
    // Commonly present optional fields
    projectId: zod_1.z.string().optional(),
    invitationId: zod_1.z.string().optional(),
    itemId: zod_1.z.string().optional(),
    itemContent: zod_1.z.string().optional(),
    responsibleUid: zod_1.z.string().optional(),
    assignedByUid: zod_1.z.string().optional(),
    fromUser: zod_1.z
        .object({
        email: zod_1.z.string(),
        fullName: zod_1.z.string(),
        id: zod_1.z.string(),
        imageId: zod_1.z.string().nullable(),
    })
        .optional(),
    projectName: zod_1.z.string().optional(),
    isDeleted: zod_1.z.boolean().optional(),
    invitationSecret: zod_1.z.string().optional(),
})
    .passthrough();
