// @doist/todoist-api-typescript/dist/cjs/types/sync/resources/reminders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReminderSchema = exports.RelativeReminderSchema = exports.AbsoluteReminderSchema = exports.LocationReminderSchema = void 0;
const zod_1 = require("zod");
const entities_1 = require("../../entities");
const ReminderBaseSchema = zod_1.z.object({
    id: zod_1.z.string(),
    notifyUid: zod_1.z.string(),
    itemId: zod_1.z.string(),
    projectId: zod_1.z.string().optional(),
    isDeleted: zod_1.z.boolean(),
});
exports.LocationReminderSchema = ReminderBaseSchema.extend({
    type: zod_1.z.literal('location'),
    name: zod_1.z.string(),
    locLat: zod_1.z.string(),
    locLong: zod_1.z.string(),
    locTrigger: zod_1.z.enum(['on_enter', 'on_leave']),
    radius: zod_1.z.number().int(),
}).passthrough();
exports.AbsoluteReminderSchema = ReminderBaseSchema.extend({
    type: zod_1.z.literal('absolute'),
    due: entities_1.DueDateSchema,
}).passthrough();
exports.RelativeReminderSchema = ReminderBaseSchema.extend({
    type: zod_1.z.literal('relative'),
    minuteOffset: zod_1.z.number().int(),
    due: entities_1.DueDateSchema.optional(),
}).passthrough();
exports.ReminderSchema = zod_1.z.discriminatedUnion('type', [
    exports.LocationReminderSchema,
    exports.AbsoluteReminderSchema,
    exports.RelativeReminderSchema,
]);
