// @doist/todoist-api-typescript/dist/cjs/types/sync/resources/suggestions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuggestionSchema = exports.WorkspaceTemplateSuggestionSchema = exports.TemplateSuggestionSchema = void 0;
const zod_1 = require("zod");
exports.TemplateSuggestionSchema = zod_1.z
    .object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    templateType: zod_1.z.enum(['project', 'setup']),
})
    .passthrough();
exports.WorkspaceTemplateSuggestionSchema = exports.TemplateSuggestionSchema.extend({
    workspaceId: zod_1.z.string().nullable(),
});
const SyncTemplateSuggestionsSchema = zod_1.z
    .object({
    type: zod_1.z.enum(['templates', 'most_used_user_templates']),
    content: zod_1.z.object({
        templates: zod_1.z.array(exports.TemplateSuggestionSchema),
        locale: zod_1.z.string(),
    }),
    isDeleted: zod_1.z.boolean(),
})
    .passthrough();
const SyncWorkspaceTemplateSuggestionsSchema = zod_1.z
    .object({
    type: zod_1.z.literal('most_used_workspace_templates'),
    content: zod_1.z.object({
        templates: zod_1.z.array(exports.WorkspaceTemplateSuggestionSchema),
        locale: zod_1.z.string(),
    }),
    isDeleted: zod_1.z.boolean(),
})
    .passthrough();
exports.SuggestionSchema = zod_1.z.union([
    SyncWorkspaceTemplateSuggestionsSchema,
    SyncTemplateSuggestionsSchema,
]);
