// @doist/todoist-api-typescript/dist/cjs/types/sync/resources/user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncUserSchema = void 0;
const zod_1 = require("zod");
const user_preferences_1 = require("../user-preferences");
const FeaturesSchema = zod_1.z
    .object({
    karmaDisabled: zod_1.z.boolean(),
    restriction: zod_1.z.number().int(),
    karmaVacation: zod_1.z.boolean(),
    dateistLang: zod_1.z.any(),
    beta: user_preferences_1.BooleanFromZeroOneSchema,
    hasPushReminders: zod_1.z.boolean(),
    dateistInlineDisabled: zod_1.z.boolean(),
    autoInviteDisabled: zod_1.z.boolean().optional(),
    goldTheme: zod_1.z.boolean().optional(),
    migratedFromTdb: zod_1.z.boolean().optional(),
})
    .passthrough();
const TzInfoSchema = zod_1.z
    .object({
    timezone: zod_1.z.string(),
    hours: zod_1.z.number().int(),
    minutes: zod_1.z.number().int(),
    isDst: user_preferences_1.BooleanFromZeroOneSchema,
    gmtString: zod_1.z.string(),
})
    .passthrough();
const JoinableWorkspaceSchema = zod_1.z
    .object({
    workspaceId: zod_1.z.string(),
    workspaceName: zod_1.z.string(),
    memberCount: zod_1.z.number().int(),
})
    .passthrough();
const GettingStartedGuideProjectSchema = zod_1.z
    .object({
    onboardingUseCase: zod_1.z.string(),
    projectId: zod_1.z.string(),
    completed: zod_1.z.boolean(),
    closed: zod_1.z.boolean(),
})
    .passthrough();
/**
 * Sync API user resource.
 *
 * This is a superset of the REST `CurrentUserSchema` — the Sync API returns
 * many additional fields for features, onboarding state, and account details.
 */
exports.SyncUserSchema = zod_1.z
    .object({
    id: zod_1.z.string(),
    email: zod_1.z.string(),
    fullName: zod_1.z.string(),
    activatedUser: zod_1.z.boolean(),
    autoReminder: zod_1.z.number().int(),
    avatarBig: zod_1.z.string().optional(),
    avatarMedium: zod_1.z.string().optional(),
    avatarS640: zod_1.z.string().optional(),
    avatarSmall: zod_1.z.string().optional(),
    businessAccountId: zod_1.z.string().nullable(),
    dailyGoal: zod_1.z.number().int(),
    dateFormat: user_preferences_1.DateFormatSchema,
    dateistLang: zod_1.z.string().nullable(),
    daysOff: zod_1.z.array(zod_1.z.number().int()),
    featureIdentifier: zod_1.z.string(),
    features: FeaturesSchema,
    freeTrailExpires: zod_1.z.string().optional(),
    hasMagicNumber: zod_1.z.boolean(),
    hasPassword: zod_1.z.boolean(),
    hasStartedATrial: zod_1.z.boolean().optional(),
    imageId: zod_1.z.string().nullable(),
    inboxProjectId: zod_1.z.string(),
    isCelebrationsEnabled: zod_1.z.boolean(),
    isPremium: zod_1.z.boolean(),
    joinableWorkspace: JoinableWorkspaceSchema.nullable(),
    joinedAt: zod_1.z.string(),
    gettingStartedGuideProjects: zod_1.z.array(GettingStartedGuideProjectSchema).nullable(),
    karma: zod_1.z.number(),
    karmaTrend: zod_1.z.string(),
    lang: zod_1.z.string(),
    mfaEnabled: zod_1.z.boolean().optional(),
    mobileHost: zod_1.z.string().nullable(),
    mobileNumber: zod_1.z.string().nullable(),
    nextWeek: user_preferences_1.DayOfWeekSchema,
    onboardingLevel: zod_1.z.string().nullable().optional(),
    onboardingRole: zod_1.z.string().nullable().optional(),
    onboardingPersona: zod_1.z.string().nullable().optional(),
    onboardingInitiated: zod_1.z.boolean().nullable().optional(),
    onboardingCompleted: zod_1.z.boolean().nullable().optional(),
    onboardingSkipped: zod_1.z.boolean().optional(),
    onboardingTeamMode: zod_1.z.boolean().nullable().optional(),
    onboardingUseCases: zod_1.z.array(zod_1.z.string()).nullable().optional(),
    premiumStatus: zod_1.z.enum([
        'not_premium',
        'current_personal_plan',
        'legacy_personal_plan',
        'teams_business_member',
    ]),
    premiumUntil: zod_1.z.string().nullable(),
    rambleSessionsUsage: zod_1.z
        .object({
        currentCount: zod_1.z.number().int(),
        limit: zod_1.z.number().int(),
        remaining: zod_1.z.number().int(),
        resetDate: zod_1.z.string(),
    })
        .nullable()
        .optional(),
    shareLimit: zod_1.z.number().int(),
    sortOrder: zod_1.z.number().int(),
    startDay: user_preferences_1.DayOfWeekSchema,
    startPage: zod_1.z.string(),
    themeId: zod_1.z.string(),
    timeFormat: user_preferences_1.TimeFormatSchema,
    token: zod_1.z.string(),
    tzInfo: TzInfoSchema,
    uniquePrefix: zod_1.z.number().int(),
    verificationStatus: zod_1.z.string(),
    websocketUrl: zod_1.z.string(),
    weekendStartDay: zod_1.z.number().int(),
    weeklyGoal: zod_1.z.number().int(),
})
    .passthrough();
