// @doist/todoist-api-typescript/dist/cjs/types/sync/resources/user-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserSettingsSchema = void 0;
const zod_1 = require("zod");
const NavigationFeatureSchema = zod_1.z.object({
    name: zod_1.z.string(),
    shown: zod_1.z.boolean(),
});
const QuickAddFeatureSchema = zod_1.z.object({
    name: zod_1.z.string(),
    shown: zod_1.z.boolean(),
});
/**
 * Sync API user settings resource.
 */
exports.UserSettingsSchema = zod_1.z
    .object({
    completedSoundDesktop: zod_1.z.boolean(),
    completedSoundMobile: zod_1.z.boolean(),
    debugLogSendingEnabledUntil: zod_1.z.string().nullable(),
    legacyPricing: zod_1.z.boolean(),
    navigation: zod_1.z
        .object({
        countsShown: zod_1.z.boolean(),
        features: zod_1.z.array(NavigationFeatureSchema),
    })
        .passthrough(),
    reminderDesktop: zod_1.z.boolean(),
    reminderEmail: zod_1.z.boolean(),
    reminderPush: zod_1.z.boolean(),
    resetRecurringSubtasks: zod_1.z.boolean(),
    aiEmailAssist: zod_1.z.boolean(),
    theme: zod_1.z.string().nullable().optional(),
    syncTheme: zod_1.z.boolean().optional(),
    quickAdd: zod_1.z
        .object({
        labelsShown: zod_1.z.boolean(),
        features: zod_1.z.array(QuickAddFeatureSchema),
    })
        .passthrough(),
})
    .passthrough();
