// @doist/todoist-api-typescript/dist/cjs/types/sync/resources/view-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectViewOptionsDefaultsSchema = exports.ViewOptionsSchema = void 0;
const zod_1 = require("zod");
const ViewTypeSchema = zod_1.z.enum([
    'TODAY',
    'UPCOMING',
    'PROJECT',
    'LABEL',
    'FILTER',
    'WORKSPACE_FILTER',
    'SEARCH',
    'TEMPLATE_PREVIEW',
    'TASK_DETAIL',
    'AUTOMATION',
    'ASSIGNED',
    'OVERDUE',
    'WORKSPACE_OVERVIEW',
]);
const ViewModeSchema = zod_1.z.enum(['LIST', 'BOARD', 'CALENDAR']);
const GroupedBySchema = zod_1.z
    .enum([
    'ASSIGNEE',
    'ADDED_DATE',
    'DUE_DATE',
    'DEADLINE',
    'LABEL',
    'PRIORITY',
    'PROJECT',
    'WORKSPACE',
])
    .nullable();
const SortedBySchema = zod_1.z
    .enum([
    'MANUAL',
    'ALPHABETICALLY',
    'ASSIGNEE',
    'DUE_DATE',
    'DEADLINE',
    'ADDED_DATE',
    'PRIORITY',
    'PROJECT',
    'WORKSPACE',
])
    .nullable();
const SortOrderSchema = zod_1.z.enum(['ASC', 'DESC']).nullable();
const CalendarSettingsSchema = zod_1.z
    .object({
    layout: zod_1.z.enum(['WEEK', 'MONTH']).optional(),
})
    .passthrough();
exports.ViewOptionsSchema = zod_1.z
    .object({
    viewType: ViewTypeSchema,
    objectId: zod_1.z.string().optional(),
    groupedBy: GroupedBySchema.optional(),
    filteredBy: zod_1.z.string().nullable().optional(),
    viewMode: ViewModeSchema.optional(),
    showCompletedTasks: zod_1.z.boolean().optional(),
    sortedBy: SortedBySchema.optional(),
    sortOrder: SortOrderSchema.optional(),
})
    .passthrough();
exports.ProjectViewOptionsDefaultsSchema = zod_1.z
    .object({
    projectId: zod_1.z.string(),
    viewMode: ViewModeSchema.nullable().optional(),
    groupedBy: GroupedBySchema.optional(),
    sortedBy: SortedBySchema.optional(),
    sortOrder: SortOrderSchema.optional(),
    showCompletedTasks: zod_1.z.boolean().optional(),
    filteredBy: zod_1.z.string().nullable().optional(),
    calendarSettings: CalendarSettingsSchema.nullable().optional(),
})
    .passthrough();
