// @doist/todoist-api-typescript/dist/cjs/types/sync/resources/workspace-goals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceGoalSchema = exports.WorkspaceGoalProgressSchema = exports.WorkspaceGoalMilestoneSchema = void 0;
const zod_1 = require("zod");
exports.WorkspaceGoalMilestoneSchema = zod_1.z
    .object({
    id: zod_1.z.string(),
    title: zod_1.z.string(),
    isCompleted: zod_1.z.boolean(),
})
    .passthrough();
exports.WorkspaceGoalProgressSchema = zod_1.z
    .object({
    completedItems: zod_1.z.number().int(),
    totalItems: zod_1.z.number().int(),
})
    .passthrough();
exports.WorkspaceGoalSchema = zod_1.z
    .object({
    id: zod_1.z.string(),
    workspaceId: zod_1.z.string(),
    title: zod_1.z.string(),
    description: zod_1.z.string().nullable(),
    deadline: zod_1.z.string().nullable(),
    isDeleted: zod_1.z.boolean(),
    projectIds: zod_1.z.array(zod_1.z.string()),
    progress: exports.WorkspaceGoalProgressSchema.nullable(),
    creatorUid: zod_1.z.string(),
    createdAt: zod_1.z.string(),
    updatedAt: zod_1.z.string(),
})
    .passthrough();
