// @doist/todoist-api-typescript/dist/cjs/types/sync/resources/workspaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncWorkspaceSchema = void 0;
const zod_1 = require("zod");
const entities_1 = require("../../entities");
/**
 * Sync API workspace resource.
 *
 * This is a superset of the REST `WorkspaceSchema` — the Sync API returns
 * additional fields like description, isDeleted, isCollapsed, member counts,
 * domain settings, and sorting preferences.
 */
exports.SyncWorkspaceSchema = zod_1.z
    .object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    logoBig: zod_1.z.string().optional(),
    logoMedium: zod_1.z.string().optional(),
    logoSmall: zod_1.z.string().optional(),
    logoS640: zod_1.z.string().optional(),
    creatorId: zod_1.z.string(),
    createdAt: zod_1.z.string(),
    isDeleted: zod_1.z.boolean(),
    isCollapsed: zod_1.z.boolean(),
    role: entities_1.WorkspaceRoleSchema,
    plan: entities_1.WorkspacePlanSchema,
    limits: entities_1.WorkspaceLimitsSchema,
    inviteCode: zod_1.z.string().nullable().optional(),
    isLinkSharingEnabled: zod_1.z.boolean().nullable().optional(),
    isGuestAllowed: zod_1.z.boolean().nullable().optional(),
    currentActiveProjects: zod_1.z.number().nullable(),
    currentMemberCount: zod_1.z.number().nullable(),
    memberCountByType: zod_1.z
        .object({
        adminCount: zod_1.z.number().int(),
        guestCount: zod_1.z.number().int(),
        memberCount: zod_1.z.number().int(),
    })
        .optional(),
    currentTemplateCount: zod_1.z.number().nullable(),
    pendingInvitations: zod_1.z.array(zod_1.z.string()).nullable().optional(),
    domainName: zod_1.z.string().optional(),
    domainDiscovery: zod_1.z.boolean().optional(),
    restrictEmailDomains: zod_1.z.boolean().optional(),
    adminSortingApplied: zod_1.z.boolean(),
    projectSortPreference: zod_1.z.string().optional(),
    defaultCollaborators: zod_1.z
        .object({
        predefinedGroupIds: zod_1.z.array(zod_1.z.string()),
        userIds: zod_1.z.array(zod_1.z.number()),
    })
        .optional(),
    properties: entities_1.WorkspacePropertiesSchema.optional(),
})
    .passthrough();
