// @doist/todoist-api-typescript/dist/cjs/types/sync/user-preferences.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DayOfWeekSchema = exports.TimeFormatSchema = exports.DateFormatSchema = exports.BooleanFromZeroOneSchema = exports.DAY_OF_WEEK_TO_API = exports.TIME_FORMAT_TO_API = exports.DATE_FORMAT_TO_API = void 0;
const zod_1 = require("zod");
const DATE_FORMAT_FROM_API = { 0: 'DD/MM/YYYY', 1: 'MM/DD/YYYY' };
const TIME_FORMAT_FROM_API = { 0: '24h', 1: '12h' };
const DAY_OF_WEEK_FROM_API = {
    1: 'Monday',
    2: 'Tuesday',
    3: 'Wednesday',
    4: 'Thursday',
    5: 'Friday',
    6: 'Saturday',
    7: 'Sunday',
};
exports.DATE_FORMAT_TO_API = {
    'DD/MM/YYYY': 0,
    'MM/DD/YYYY': 1,
};
exports.TIME_FORMAT_TO_API = { '24h': 0, '12h': 1 };
exports.DAY_OF_WEEK_TO_API = {
    Monday: 1,
    Tuesday: 2,
    Wednesday: 3,
    Thursday: 4,
    Friday: 5,
    Saturday: 6,
    Sunday: 7,
};
/** Zod read-schema: parse API 0/1 integer, emit boolean */
exports.BooleanFromZeroOneSchema = zod_1.z
    .union([zod_1.z.literal(0), zod_1.z.literal(1)])
    .transform((v) => v === 1);
/** Zod read-schemas: parse API numbers, emit descriptive strings */
exports.DateFormatSchema = zod_1.z
    .union([zod_1.z.literal(0), zod_1.z.literal(1)])
    .transform((v) => DATE_FORMAT_FROM_API[v]);
exports.TimeFormatSchema = zod_1.z
    .union([zod_1.z.literal(0), zod_1.z.literal(1)])
    .transform((v) => TIME_FORMAT_FROM_API[v]);
exports.DayOfWeekSchema = zod_1.z
    .union([
    zod_1.z.literal(1),
    zod_1.z.literal(2),
    zod_1.z.literal(3),
    zod_1.z.literal(4),
    zod_1.z.literal(5),
    zod_1.z.literal(6),
    zod_1.z.literal(7),
])
    .transform((v) => DAY_OF_WEEK_FROM_API[v]);
