// @doist/todoist-api-typescript/dist/cjs/utils/colors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultColor = exports.colors = exports.taupe = exports.grey = exports.charcoal = exports.salmon = exports.magenta = exports.lavender = exports.violet = exports.grape = exports.blue = exports.lightBlue = exports.skyBlue = exports.teal = exports.mintGreen = exports.green = exports.limeGreen = exports.oliveGreen = exports.yellow = exports.orange = exports.red = exports.berryRed = void 0;
exports.getColorByKey = getColorByKey;
exports.berryRed = {
    key: 'berry_red',
    displayName: 'Berry Red',
    hexValue: '#b8255f',
};
exports.red = {
    key: 'red',
    displayName: 'Red',
    hexValue: '#db4035',
};
exports.orange = {
    key: 'orange',
    displayName: 'Orange',
    hexValue: '#ff9933',
};
exports.yellow = {
    key: 'yellow',
    displayName: 'Yellow',
    hexValue: '#fad000',
};
exports.oliveGreen = {
    key: 'olive_green',
    displayName: 'Olive Green',
    hexValue: '#afb83b',
};
exports.limeGreen = {
    key: 'lime_green',
    displayName: 'Lime Green',
    hexValue: '#7ecc49',
};
exports.green = {
    key: 'green',
    displayName: 'Green',
    hexValue: '#299438',
};
exports.mintGreen = {
    key: 'mint_green',
    displayName: 'Mint Green',
    hexValue: '#6accbc',
};
exports.teal = {
    key: 'teal',
    displayName: 'Teal',
    hexValue: '#158fad',
};
exports.skyBlue = {
    key: 'sky_blue',
    displayName: 'Sky Blue',
    hexValue: '#14aaf5',
};
exports.lightBlue = {
    key: 'light_blue',
    displayName: 'Light Blue',
    hexValue: '#96c3eb',
};
exports.blue = {
    key: 'blue',
    displayName: 'Blue',
    hexValue: '#4073ff',
};
exports.grape = {
    key: 'grape',
    displayName: 'Grape',
    hexValue: '#884dff',
};
exports.violet = {
    key: 'violet',
    displayName: 'Violet',
    hexValue: '#af38eb',
};
exports.lavender = {
    key: 'lavender',
    displayName: 'Lavender',
    hexValue: '#eb96eb',
};
exports.magenta = {
    key: 'magenta',
    displayName: 'Magenta',
    hexValue: '#e05194',
};
exports.salmon = {
    key: 'salmon',
    displayName: 'Salmon',
    hexValue: '#ff8d85',
};
exports.charcoal = {
    key: 'charcoal',
    displayName: 'Charcoal',
    hexValue: '#808080',
};
exports.grey = {
    key: 'grey',
    displayName: 'Grey',
    hexValue: '#b8b8b8',
};
exports.taupe = {
    key: 'taupe',
    displayName: 'Taupe',
    hexValue: '#ccac93',
};
exports.colors = [
    exports.berryRed,
    exports.red,
    exports.orange,
    exports.yellow,
    exports.oliveGreen,
    exports.limeGreen,
    exports.green,
    exports.mintGreen,
    exports.teal,
    exports.skyBlue,
    exports.lightBlue,
    exports.blue,
    exports.grape,
    exports.violet,
    exports.lavender,
    exports.magenta,
    exports.salmon,
    exports.charcoal,
    exports.grey,
    exports.taupe,
];
exports.defaultColor = exports.charcoal;
/**
 * Retrieves a {@link Color} object by its key identifier.
 *
 * @param colorKey - The unique key identifier of the color to find (e.g., 'berry_red', 'sky_blue')
 * @returns The matching Color object if found, otherwise returns the default color (charcoal)
 *
 * @example
 * ```typescript
 * const color = getColorByKey('berry_red');
 * console.log(color.hexValue); // '#b8255f'
 * ```
 */
function getColorByKey(colorKey) {
    const color = exports.colors.find((color) => color.key === colorKey);
    return color !== null && color !== void 0 ? color : exports.defaultColor;
}
