// @doist/todoist-api-typescript/dist/cjs/utils/multipart-upload.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.uploadMultipartFile = uploadMultipartFile;
const fetch_with_retry_1 = require("./fetch-with-retry");
/**
 * Helper function to determine content-type from filename extension.
 * @param fileName - The filename to analyze
 * @returns The appropriate MIME type
 */
function getContentTypeFromFileName(fileName) {
    const extension = fileName.toLowerCase().split('.').pop();
    switch (extension) {
        case 'png':
            return 'image/png';
        case 'jpg':
        case 'jpeg':
            return 'image/jpeg';
        case 'gif':
            return 'image/gif';
        case 'webp':
            return 'image/webp';
        case 'svg':
            return 'image/svg+xml';
        default:
            return 'application/octet-stream';
    }
}
/**
 * Uploads a file using multipart/form-data.
 *
 * This is a shared utility for uploading files to Todoist endpoints that require
 * multipart/form-data content type (e.g., file uploads, workspace logo uploads).
 *
 * Supports both browser (Blob/File) and Node.js (Buffer/ReadableStream/path) environments.
 *
 * @param baseUrl - The base API URL (e.g., https://api.todoist.com/api/v1/)
 * @param authToken - The authentication token
 * @param endpoint - The relative endpoint path (e.g., 'uploads', 'workspaces/logo')
 * @param file - The file content (Blob/File for browser, or Buffer/ReadableStream/path for Node)
 * @param fileName - Optional file name (required for Buffer/Stream, optional for paths and File objects)
 * @param additionalFields - Additional form fields to include (e.g., project_id, workspace_id)
 * @param requestId - Optional request ID for idempotency
 * @returns The response data from the server
 *
 * @example
 * ```typescript
 * // Upload from a file path
 * const result = await uploadMultipartFile(
 *   'https://api.todoist.com/api/v1/',
 *   'my-token',
 *   'uploads',
 *   '/path/to/file.pdf',
 *   undefined,
 *   { project_id: '12345' }
 * )
 *
 * // Upload from a Buffer
 * const buffer = Buffer.from('file content')
 * const result = await uploadMultipartFile(
 *   'https://api.todoist.com/api/v1/',
 *   'my-token',
 *   'uploads',
 *   buffer,
 *   'document.pdf',
 *   { project_id: '12345' }
 * )
 * ```
 */
async function uploadMultipartFile(args) {
    const { baseUrl, authToken, endpoint, file, fileName, additionalFields, requestId, customFetch, } = args;
    // Build the full URL
    const url = `${baseUrl}${endpoint}`;
    let body;
    const headers = {
        Authorization: `Bearer ${authToken}`,
    };
    if (requestId) {
        headers['X-Request-Id'] = requestId;
    }
    if (file instanceof Blob) {
        // Browser path: use native FormData
        const form = new globalThis.FormData();
        const resolvedFileName = fileName || (file instanceof File ? file.name : undefined) || 'upload';
        form.append('file', file, resolvedFileName);
        for (const [key, value] of Object.entries(additionalFields)) {
            if (value !== undefined && value !== null) {
                form.append(key, value.toString());
            }
        }
        // Don't set Content-Type — let fetch set it with the correct multipart boundary
        body = form;
    }
    else {
        // Node path: dynamically import Node-only modules
        const [FormDataModule, fsModule, pathModule] = await Promise.all([
            Promise.resolve().then(() => __importStar(require('form-data'))),
            Promise.resolve().then(() => __importStar(require('fs'))),
            Promise.resolve().then(() => __importStar(require('path'))),
        ]);
        const FormData = FormDataModule.default;
        const form = new FormData();
        if (typeof file === 'string') {
            // File path - create read stream
            const resolvedFileName = fileName || pathModule.basename(file);
            form.append('file', fsModule.createReadStream(file), resolvedFileName);
        }
        else if (Buffer.isBuffer(file)) {
            // Buffer - require fileName
            if (!fileName) {
                throw new Error('fileName is required when uploading from a Buffer');
            }
            const contentType = getContentTypeFromFileName(fileName);
            form.append('file', file, {
                filename: fileName,
                contentType: contentType,
            });
        }
        else {
            // Stream - require fileName
            if (!fileName) {
                throw new Error('fileName is required when uploading from a stream');
            }
            form.append('file', file, fileName);
        }
        for (const [key, value] of Object.entries(additionalFields)) {
            if (value !== undefined && value !== null) {
                form.append(key, value.toString());
            }
        }
        Object.assign(headers, form.getHeaders());
        body = form;
    }
    // Make the request using fetch
    const response = await (0, fetch_with_retry_1.fetchWithRetry)({
        url,
        options: {
            method: 'POST',
            body,
            headers,
            timeout: 30000, // 30 second timeout for file uploads
        },
        customFetch,
    });
    return response.data;
}
