// @doist/todoist-api-typescript/dist/cjs/utils/validators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateUserPlanLimits = exports.validateProjectViewOptionsDefaultsArray = exports.validateProjectViewOptionsDefaults = exports.validateViewOptionsArray = exports.validateViewOptions = exports.validateCompletedInfoArray = exports.validateCompletedInfo = exports.validateReminderArray = exports.validateReminder = exports.validateCalendarAccountArray = exports.validateCalendarAccount = exports.validateCalendarArray = exports.validateCalendar = exports.validateWorkspaceGoalArray = exports.validateWorkspaceGoal = exports.validateWorkspaceFilterArray = exports.validateWorkspaceFilter = exports.validateTooltips = exports.validateNoteArray = exports.validateNote = exports.validateFolderArray = exports.validateFolder = exports.validateCollaboratorStateArray = exports.validateCollaboratorState = exports.validateCollaboratorArray = exports.validateCollaborator = exports.validateFilterArray = exports.validateFilter = exports.validateWorkspaceArray = exports.validateWorkspace = exports.validateJoinWorkspaceResult = exports.validateWorkspacePlanDetails = exports.validateWorkspaceInvitationArray = exports.validateWorkspaceInvitation = exports.validateWorkspaceUser = exports.validateAttachment = exports.validateActivityEventArray = exports.validateActivityEvent = exports.validateCurrentUser = exports.validateProductivityStats = exports.validateUserArray = exports.validateUser = exports.validateCommentArray = exports.validateComment = exports.validateLabelArray = exports.validateLabel = exports.validateSectionArray = exports.validateSection = exports.validateTaskArray = exports.validateTask = void 0;
exports.validateSuggestionArray = exports.validateSuggestion = exports.validateUserSettings = exports.validateSyncUser = exports.validateSyncWorkspaceArray = exports.validateSyncWorkspace = exports.validateLiveNotificationArray = exports.validateLiveNotification = void 0;
exports.isWorkspaceProject = isWorkspaceProject;
exports.isPersonalProject = isPersonalProject;
exports.validateProject = validateProject;
exports.validateProjectArray = validateProjectArray;
exports.validateWorkspaceUserArray = validateWorkspaceUserArray;
const entities_1 = require("../types/entities");
const resources_1 = require("../types/sync/resources");
function createValidator(schema) {
    return (input) => schema.parse(input);
}
function createArrayValidator(validateItem) {
    return (input) => input.map(validateItem);
}
// Entity validators
exports.validateTask = createValidator(entities_1.TaskSchema);
exports.validateTaskArray = createArrayValidator(exports.validateTask);
/**
 * Type guard to check if a project is a workspace project.
 * @param project The project to check
 * @returns True if the project is a workspace project
 */
function isWorkspaceProject(project) {
    return 'workspaceId' in project;
}
/**
 * Type guard to check if a project is a personal project.
 * @param project The project to check
 * @returns True if the project is a personal project
 */
function isPersonalProject(project) {
    return !isWorkspaceProject(project);
}
/**
 * Validates and parses a project input.
 * @param input The input to validate
 * @returns A validated project (either PersonalProject or WorkspaceProject)
 */
function validateProject(input) {
    if ('workspaceId' in input) {
        return entities_1.WorkspaceProjectSchema.parse(input);
    }
    return entities_1.PersonalProjectSchema.parse(input);
}
function validateProjectArray(input) {
    return input.map(validateProject);
}
exports.validateSection = createValidator(entities_1.SectionSchema);
exports.validateSectionArray = createArrayValidator(exports.validateSection);
exports.validateLabel = createValidator(entities_1.LabelSchema);
exports.validateLabelArray = createArrayValidator(exports.validateLabel);
exports.validateComment = createValidator(entities_1.CommentSchema);
exports.validateCommentArray = createArrayValidator(exports.validateComment);
exports.validateUser = createValidator(entities_1.UserSchema);
exports.validateUserArray = createArrayValidator(exports.validateUser);
exports.validateProductivityStats = createValidator(entities_1.ProductivityStatsSchema);
exports.validateCurrentUser = createValidator(entities_1.CurrentUserSchema);
exports.validateActivityEvent = createValidator(entities_1.ActivityEventSchema);
exports.validateActivityEventArray = createArrayValidator(exports.validateActivityEvent);
exports.validateAttachment = createValidator(entities_1.AttachmentSchema);
exports.validateWorkspaceUser = createValidator(entities_1.WorkspaceUserSchema);
function validateWorkspaceUserArray(input) {
    if (!Array.isArray(input)) {
        throw new Error(`Expected array for workspace users, got ${typeof input}`);
    }
    return input.map(exports.validateWorkspaceUser);
}
exports.validateWorkspaceInvitation = createValidator(entities_1.WorkspaceInvitationSchema);
exports.validateWorkspaceInvitationArray = createArrayValidator(exports.validateWorkspaceInvitation);
exports.validateWorkspacePlanDetails = createValidator(entities_1.WorkspacePlanDetailsSchema);
exports.validateJoinWorkspaceResult = createValidator(entities_1.JoinWorkspaceResultSchema);
exports.validateWorkspace = createValidator(entities_1.WorkspaceSchema);
exports.validateWorkspaceArray = createArrayValidator(exports.validateWorkspace);
// Sync resource validators
exports.validateFilter = createValidator(resources_1.FilterSchema);
exports.validateFilterArray = createArrayValidator(exports.validateFilter);
exports.validateCollaborator = createValidator(resources_1.CollaboratorSchema);
exports.validateCollaboratorArray = createArrayValidator(exports.validateCollaborator);
exports.validateCollaboratorState = createValidator(resources_1.CollaboratorStateSchema);
exports.validateCollaboratorStateArray = createArrayValidator(exports.validateCollaboratorState);
exports.validateFolder = createValidator(resources_1.FolderSchema);
exports.validateFolderArray = createArrayValidator(exports.validateFolder);
exports.validateNote = createValidator(resources_1.NoteSchema);
exports.validateNoteArray = createArrayValidator(exports.validateNote);
exports.validateTooltips = createValidator(resources_1.TooltipsSchema);
exports.validateWorkspaceFilter = createValidator(resources_1.WorkspaceFilterSchema);
exports.validateWorkspaceFilterArray = createArrayValidator(exports.validateWorkspaceFilter);
exports.validateWorkspaceGoal = createValidator(resources_1.WorkspaceGoalSchema);
exports.validateWorkspaceGoalArray = createArrayValidator(exports.validateWorkspaceGoal);
exports.validateCalendar = createValidator(resources_1.CalendarSchema);
exports.validateCalendarArray = createArrayValidator(exports.validateCalendar);
exports.validateCalendarAccount = createValidator(resources_1.CalendarAccountSchema);
exports.validateCalendarAccountArray = createArrayValidator(exports.validateCalendarAccount);
exports.validateReminder = createValidator(resources_1.ReminderSchema);
exports.validateReminderArray = createArrayValidator(exports.validateReminder);
exports.validateCompletedInfo = createValidator(resources_1.CompletedInfoSchema);
exports.validateCompletedInfoArray = createArrayValidator(exports.validateCompletedInfo);
exports.validateViewOptions = createValidator(resources_1.ViewOptionsSchema);
exports.validateViewOptionsArray = createArrayValidator(exports.validateViewOptions);
exports.validateProjectViewOptionsDefaults = createValidator(resources_1.ProjectViewOptionsDefaultsSchema);
exports.validateProjectViewOptionsDefaultsArray = createArrayValidator(exports.validateProjectViewOptionsDefaults);
exports.validateUserPlanLimits = createValidator(resources_1.UserPlanLimitsSchema);
exports.validateLiveNotification = createValidator(resources_1.LiveNotificationSchema);
exports.validateLiveNotificationArray = createArrayValidator(exports.validateLiveNotification);
exports.validateSyncWorkspace = createValidator(resources_1.SyncWorkspaceSchema);
exports.validateSyncWorkspaceArray = createArrayValidator(exports.validateSyncWorkspace);
exports.validateSyncUser = createValidator(resources_1.SyncUserSchema);
exports.validateUserSettings = createValidator(resources_1.UserSettingsSchema);
exports.validateSuggestion = createValidator(resources_1.SuggestionSchema);
exports.validateSuggestionArray = createArrayValidator(exports.validateSuggestion);
