// @doist/todoist-api-typescript/dist/cjs/test-utils/msw-setup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpResponse = exports.http = exports.server = exports.handlers = void 0;
exports.captureRequest = captureRequest;
exports.getLastRequest = getLastRequest;
exports.getAllRequests = getAllRequests;
exports.clearCapturedRequests = clearCapturedRequests;
exports.mockApiResponse = mockApiResponse;
exports.mockApiError = mockApiError;
const node_1 = require("msw/node");
const msw_1 = require("msw");
Object.defineProperty(exports, "http", { enumerable: true, get: function () { return msw_1.http; } });
Object.defineProperty(exports, "HttpResponse", { enumerable: true, get: function () { return msw_1.HttpResponse; } });
// Request capture storage
const capturedRequests = [];
// Helper function to capture requests
function captureRequest({ request, body }) {
    const headers = {};
    request.headers.forEach((value, key) => {
        headers[key] = value;
    });
    capturedRequests.push({
        url: request.url,
        method: request.method,
        headers,
        body,
    });
}
// Helper function to get the last captured request
function getLastRequest() {
    return capturedRequests[capturedRequests.length - 1];
}
// Helper function to get all captured requests
function getAllRequests() {
    return [...capturedRequests];
}
// Helper function to clear captured requests
function clearCapturedRequests() {
    capturedRequests.length = 0;
}
// Default handlers for common API responses
exports.handlers = [
// Default handlers can be added here for common endpoints
// Individual test files will add their own specific handlers
];
// Create MSW server instance
exports.server = (0, node_1.setupServer)(...exports.handlers);
// Helper to create a resolver function for MSW handlers
function createResolver(data, status, headers) {
    return async function resolver({ request }) {
        let body = undefined;
        if (request.method !== 'GET') {
            try {
                body = await request.json();
            }
            catch (_a) {
                // Body might not be JSON
                try {
                    body = await request.text();
                }
                catch (_b) {
                    // Body might be FormData or not parseable
                }
            }
        }
        captureRequest({ request, body });
        return msw_1.HttpResponse.json(data, {
            status,
            headers,
        });
    };
}
// Helper to mock a successful API response
function mockApiResponse({ endpoint, data, options = {}, }) {
    const { status = 200, method = 'GET', headers = {} } = options;
    const resolver = createResolver(data, status, headers);
    const httpMethod = method.toLowerCase();
    exports.server.use(msw_1.http[httpMethod](endpoint, resolver));
}
// Helper to mock an error response
function mockApiError({ endpoint, data, status, options = {}, }) {
    mockApiResponse({ endpoint, data, options: Object.assign(Object.assign({}, options), { status }) });
}
// Setup MSW for tests
beforeAll(() => {
    exports.server.listen({
        onUnhandledRequest: 'error', // Throw errors for unhandled requests to catch unexpected fetch calls
    });
});
afterEach(() => {
    exports.server.resetHandlers(); // Reset handlers between tests
    clearCapturedRequests(); // Clear captured requests between tests
});
afterAll(() => {
    exports.server.close(); // Clean up after all tests
});
