// @doist/todoist-api-typescript/dist/cjs/test-utils/test-defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = exports.COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = exports.RAW_COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = exports.COMMENT_WITH_OPTIONALS_AS_NULL_TASK = exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_TASK = exports.INVALID_COMMENT = exports.DEFAULT_COMMENT = exports.DEFAULT_RAW_COMMENT = exports.INVALID_ATTACHMENT = exports.DEFAULT_ATTACHMENT = exports.INVALID_USER = exports.DEFAULT_USER = exports.INVALID_LABEL = exports.DEFAULT_LABEL = exports.INVALID_SECTION = exports.DEFAULT_SECTION = exports.PROJECT_WITH_OPTIONALS_AS_NULL = exports.INVALID_PROJECT = exports.DEFAULT_PROJECT = exports.TASK_WITH_OPTIONALS_AS_NULL = exports.INVALID_TASK = exports.DEFAULT_TASK = exports.DEFAULT_DEADLINE = exports.DEFAULT_DURATION = exports.DEFAULT_DUE_DATE = exports.INVALID_ENTITY_ID = exports.DEFAULT_REQUEST_ID = exports.DEFAULT_AUTH_TOKEN = exports.DEFAULT_PROJECT_VIEW_STYLE = exports.DEFAULT_PROJECT_NAME = exports.DEFAULT_PROJECT_ID = exports.DEFAULT_ORDER = exports.DEFAULT_TASK_PRIORITY = exports.DEFAULT_TASK_DESCRIPTION = exports.DEFAULT_TASK_CONTENT = exports.DEFAULT_TASK_ID = void 0;
const url_helpers_1 = require("../utils/url-helpers");
exports.DEFAULT_TASK_ID = '1234';
exports.DEFAULT_TASK_CONTENT = 'This is a task';
exports.DEFAULT_TASK_DESCRIPTION = 'A description';
exports.DEFAULT_TASK_PRIORITY = 1;
exports.DEFAULT_ORDER = 3;
exports.DEFAULT_PROJECT_ID = '123';
exports.DEFAULT_PROJECT_NAME = 'This is a project';
exports.DEFAULT_PROJECT_VIEW_STYLE = 'list';
const DEFAULT_LABEL_ID = '456';
const DEFAULT_LABEL_NAME = 'This is a label';
const DEFAULT_SECTION_ID = '456';
const DEFAULT_SECTION_NAME = 'This is a section';
const DEFAULT_PARENT_ID = '5678';
const DEFAULT_ASSIGNEE = '1234';
const DEFAULT_CREATOR = '1234';
const DEFAULT_DATE = '2020-09-08T12:00:00Z';
const DEFAULT_ENTITY_COLOR = 'berry_red';
const DEFAULT_LABELS = ['personal', 'work', 'hobby'];
const DEFAULT_USER_ID = '5';
const DEFAULT_USER_NAME = 'A User';
const DEFAULT_USER_EMAIL = 'atestuser@doist.com';
const DEFAULT_COMMENT_ID = '4';
const DEFAULT_COMMENT_CONTENT = 'A comment';
const DEFAULT_COMMENT_REACTIONS = { '👍': ['1234', '5678'] };
const DEFAULT_CAN_ASSIGN_TASKS = true;
const DEFAULT_IS_ARCHIVED = false;
const DEFAULT_IS_DELETED = false;
const DEFAULT_IS_FROZEN = false;
const DEFAULT_IS_COLLAPSED = false;
// URL constants using the helper functions
const DEFAULT_TASK_URL = (0, url_helpers_1.getTaskUrl)(exports.DEFAULT_TASK_ID, exports.DEFAULT_TASK_CONTENT);
const DEFAULT_PROJECT_URL = (0, url_helpers_1.getProjectUrl)(exports.DEFAULT_PROJECT_ID, exports.DEFAULT_PROJECT_NAME);
const DEFAULT_SECTION_URL = (0, url_helpers_1.getSectionUrl)(DEFAULT_SECTION_ID, DEFAULT_SECTION_NAME);
exports.DEFAULT_AUTH_TOKEN = 'AToken';
exports.DEFAULT_REQUEST_ID = 'ARequestID';
exports.INVALID_ENTITY_ID = 1234;
exports.DEFAULT_DUE_DATE = {
    isRecurring: false,
    string: 'a date string',
    date: DEFAULT_DATE,
    lang: 'en',
    timezone: null,
};
exports.DEFAULT_DURATION = {
    amount: 10,
    unit: 'minute',
};
exports.DEFAULT_DEADLINE = {
    date: '2020-09-08',
    lang: 'en',
};
exports.DEFAULT_TASK = {
    id: exports.DEFAULT_TASK_ID,
    userId: DEFAULT_CREATOR,
    projectId: exports.DEFAULT_PROJECT_ID,
    sectionId: DEFAULT_SECTION_ID,
    parentId: DEFAULT_PARENT_ID,
    addedByUid: DEFAULT_CREATOR,
    assignedByUid: DEFAULT_CREATOR,
    responsibleUid: DEFAULT_ASSIGNEE,
    labels: DEFAULT_LABELS,
    deadline: exports.DEFAULT_DEADLINE,
    duration: exports.DEFAULT_DURATION,
    checked: false,
    isDeleted: DEFAULT_IS_DELETED,
    addedAt: DEFAULT_DATE,
    completedAt: null,
    updatedAt: DEFAULT_DATE,
    due: exports.DEFAULT_DUE_DATE,
    priority: exports.DEFAULT_TASK_PRIORITY,
    childOrder: exports.DEFAULT_ORDER,
    content: exports.DEFAULT_TASK_CONTENT,
    description: exports.DEFAULT_TASK_DESCRIPTION,
    dayOrder: exports.DEFAULT_ORDER,
    isCollapsed: DEFAULT_IS_COLLAPSED,
    isUncompletable: false,
    url: DEFAULT_TASK_URL,
};
exports.INVALID_TASK = Object.assign(Object.assign({}, exports.DEFAULT_TASK), { due: '2020-01-31' });
exports.TASK_WITH_OPTIONALS_AS_NULL = {
    userId: DEFAULT_CREATOR,
    id: exports.DEFAULT_TASK_ID,
    projectId: exports.DEFAULT_PROJECT_ID,
    sectionId: null,
    parentId: null,
    addedByUid: DEFAULT_CREATOR,
    assignedByUid: null,
    responsibleUid: null,
    labels: [],
    deadline: null,
    duration: null,
    checked: false,
    isDeleted: DEFAULT_IS_DELETED,
    addedAt: DEFAULT_DATE,
    completedAt: null,
    updatedAt: DEFAULT_DATE,
    due: null,
    priority: exports.DEFAULT_TASK_PRIORITY,
    childOrder: exports.DEFAULT_ORDER,
    content: exports.DEFAULT_TASK_CONTENT,
    description: exports.DEFAULT_TASK_DESCRIPTION,
    dayOrder: exports.DEFAULT_ORDER,
    isCollapsed: DEFAULT_IS_COLLAPSED,
    isUncompletable: false,
    url: DEFAULT_TASK_URL,
};
exports.DEFAULT_PROJECT = {
    id: exports.DEFAULT_PROJECT_ID,
    name: exports.DEFAULT_PROJECT_NAME,
    color: DEFAULT_ENTITY_COLOR,
    childOrder: exports.DEFAULT_ORDER,
    parentId: exports.DEFAULT_PROJECT_ID,
    isFavorite: false,
    isShared: false,
    inboxProject: false,
    viewStyle: exports.DEFAULT_PROJECT_VIEW_STYLE,
    canAssignTasks: DEFAULT_CAN_ASSIGN_TASKS,
    isArchived: DEFAULT_IS_ARCHIVED,
    isDeleted: DEFAULT_IS_DELETED,
    isFrozen: DEFAULT_IS_FROZEN,
    createdAt: DEFAULT_DATE,
    updatedAt: DEFAULT_DATE,
    defaultOrder: exports.DEFAULT_ORDER,
    description: '',
    isCollapsed: DEFAULT_IS_COLLAPSED,
    url: DEFAULT_PROJECT_URL,
};
exports.INVALID_PROJECT = Object.assign(Object.assign({}, exports.DEFAULT_PROJECT), { name: 123 });
exports.PROJECT_WITH_OPTIONALS_AS_NULL = Object.assign(Object.assign({}, exports.DEFAULT_PROJECT), { parentId: null });
exports.DEFAULT_SECTION = {
    id: DEFAULT_SECTION_ID,
    userId: DEFAULT_USER_ID,
    projectId: exports.DEFAULT_PROJECT_ID,
    addedAt: '2025-03-28T14:01:23.334881Z',
    updatedAt: '2025-03-28T14:01:23.334885Z',
    archivedAt: null,
    name: DEFAULT_SECTION_NAME,
    sectionOrder: exports.DEFAULT_ORDER,
    isArchived: false,
    isDeleted: false,
    isCollapsed: false,
    url: DEFAULT_SECTION_URL,
};
exports.INVALID_SECTION = Object.assign(Object.assign({}, exports.DEFAULT_SECTION), { projectId: undefined });
exports.DEFAULT_LABEL = {
    id: DEFAULT_LABEL_ID,
    name: DEFAULT_LABEL_NAME,
    color: DEFAULT_ENTITY_COLOR,
    order: exports.DEFAULT_ORDER,
    isFavorite: false,
};
exports.INVALID_LABEL = Object.assign(Object.assign({}, exports.DEFAULT_LABEL), { isFavorite: 'true' });
exports.DEFAULT_USER = {
    id: DEFAULT_USER_ID,
    name: DEFAULT_USER_NAME,
    email: DEFAULT_USER_EMAIL,
};
exports.INVALID_USER = Object.assign(Object.assign({}, exports.DEFAULT_USER), { email: undefined });
exports.DEFAULT_ATTACHMENT = {
    resourceType: 'file',
    fileType: 'image/png',
    fileUrl: 'https://someurl.com/image.jpg',
    uploadState: 'completed',
};
exports.INVALID_ATTACHMENT = Object.assign(Object.assign({}, exports.DEFAULT_ATTACHMENT), { uploadState: 'something random' });
exports.DEFAULT_RAW_COMMENT = {
    id: DEFAULT_COMMENT_ID,
    postedUid: DEFAULT_USER_ID,
    content: DEFAULT_COMMENT_CONTENT,
    fileAttachment: exports.DEFAULT_ATTACHMENT,
    uidsToNotify: null,
    isDeleted: false,
    postedAt: DEFAULT_DATE,
    reactions: DEFAULT_COMMENT_REACTIONS,
    itemId: exports.DEFAULT_TASK_ID,
};
exports.DEFAULT_COMMENT = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { taskId: exports.DEFAULT_RAW_COMMENT.itemId, itemId: undefined });
exports.INVALID_COMMENT = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { isDeleted: 'true' });
exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_TASK = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { fileAttachment: null, uidsToNotify: null, reactions: null });
exports.COMMENT_WITH_OPTIONALS_AS_NULL_TASK = Object.assign(Object.assign({}, exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_TASK), { taskId: exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_TASK.itemId, itemId: undefined });
exports.RAW_COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { fileAttachment: {
        resourceType: 'file',
        fileName: null,
        fileSize: null,
        fileType: null,
        fileDuration: null,
        uploadState: null,
        image: null,
        imageWidth: null,
        imageHeight: null,
        url: null,
        title: null,
    } });
exports.COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = Object.assign(Object.assign({}, exports.RAW_COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL), { taskId: exports.RAW_COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL.itemId, itemId: undefined });
exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { itemId: undefined, projectId: exports.DEFAULT_PROJECT_ID });
exports.COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = Object.assign(Object.assign({}, exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT), { taskId: undefined });
