// @doist/todoist-api-typescript/dist/esm/test-utils/msw-setup.js
import { setupServer } from 'msw/node';
import { http, HttpResponse } from 'msw';
// Request capture storage
const capturedRequests = [];
// Helper function to capture requests
export function captureRequest({ request, body }) {
    const headers = {};
    request.headers.forEach((value, key) => {
        headers[key] = value;
    });
    capturedRequests.push({
        url: request.url,
        method: request.method,
        headers,
        body,
    });
}
// Helper function to get the last captured request
export function getLastRequest() {
    return capturedRequests[capturedRequests.length - 1];
}
// Helper function to get all captured requests
export function getAllRequests() {
    return [...capturedRequests];
}
// Helper function to clear captured requests
export function clearCapturedRequests() {
    capturedRequests.length = 0;
}
// Default handlers for common API responses
export const handlers = [
// Default handlers can be added here for common endpoints
// Individual test files will add their own specific handlers
];
// Create MSW server instance
export const server = setupServer(...handlers);
// Helper to create a resolver function for MSW handlers
function createResolver(data, status, headers) {
    return async function resolver({ request }) {
        let body = undefined;
        if (request.method !== 'GET') {
            try {
                body = await request.json();
            }
            catch (_a) {
                // Body might not be JSON
                try {
                    body = await request.text();
                }
                catch (_b) {
                    // Body might be FormData or not parseable
                }
            }
        }
        captureRequest({ request, body });
        return HttpResponse.json(data, {
            status,
            headers,
        });
    };
}
// Helper to mock a successful API response
export function mockApiResponse({ endpoint, data, options = {}, }) {
    const { status = 200, method = 'GET', headers = {} } = options;
    const resolver = createResolver(data, status, headers);
    const httpMethod = method.toLowerCase();
    server.use(http[httpMethod](endpoint, resolver));
}
// Helper to mock an error response
export function mockApiError({ endpoint, data, status, options = {}, }) {
    mockApiResponse({ endpoint, data, options: Object.assign(Object.assign({}, options), { status }) });
}
// Setup MSW for tests
beforeAll(() => {
    server.listen({
        onUnhandledRequest: 'error', // Throw errors for unhandled requests to catch unexpected fetch calls
    });
});
afterEach(() => {
    server.resetHandlers(); // Reset handlers between tests
    clearCapturedRequests(); // Clear captured requests between tests
});
afterAll(() => {
    server.close(); // Clean up after all tests
});
// Export MSW utilities for use in tests
export { http, HttpResponse };
