// @doist/todoist-ai/dist/index.js
import { f as s, s as a, l as e, m as t, a as d, u as o, d as n, b as i, c as r, e as c, h as m, i as f, j as p, k as l, n as k, o as T, p as u, q as j, r as C, t as b, v as g, w as v, x as A, y as S, z as h, A as P, B as x, C as y } from "./mcp-server-DqAX_RqB.js";
import { F as L, g as M } from "./mcp-server-DqAX_RqB.js";
const w = {
  // Task management tools
  addTasks: y,
  completeTasks: x,
  uncompleteTasks: P,
  updateTasks: h,
  findTasks: S,
  findTasksByDate: A,
  findCompletedTasks: v,
  rescheduleTasks: g,
  // Project management tools
  addProjects: b,
  updateProjects: C,
  findProjects: j,
  // Section management tools
  addSections: u,
  updateSections: T,
  findSections: k,
  // Comment management tools
  addComments: l,
  updateComments: p,
  findComments: f,
  // Label management tools
  addLabels: m,
  findLabels: c,
  // Activity and audit tools
  findActivity: r,
  // General tools
  getOverview: i,
  deleteObject: n,
  userInfo: o,
  // Assignment and collaboration tools
  findProjectCollaborators: d,
  manageAssignments: t,
  // Workspace tools
  listWorkspaces: e,
  // OpenAI MCP tools
  search: a,
  fetch: s
};
export {
  L as FEATURE_NAMES,
  l as addComments,
  m as addLabels,
  b as addProjects,
  u as addSections,
  y as addTasks,
  x as completeTasks,
  n as deleteObject,
  s as fetch,
  r as findActivity,
  f as findComments,
  v as findCompletedTasks,
  c as findLabels,
  d as findProjectCollaborators,
  j as findProjects,
  k as findSections,
  S as findTasks,
  A as findTasksByDate,
  M as getMcpServer,
  i as getOverview,
  e as listWorkspaces,
  t as manageAssignments,
  g as rescheduleTasks,
  a as search,
  w as tools,
  P as uncompleteTasks,
  p as updateComments,
  C as updateProjects,
  T as updateSections,
  h as updateTasks,
  o as userInfo
};
