// @doist/todoist-api-typescript/dist/esm/types/sync/resources/reminders.js
import { z } from 'zod';
import { DueDateSchema } from '../../entities.js';
const ReminderBaseSchema = z.object({
    id: z.string(),
    notifyUid: z.string(),
    itemId: z.string(),
    projectId: z.string().optional(),
    isDeleted: z.boolean(),
});
export const LocationReminderSchema = ReminderBaseSchema.extend({
    type: z.literal('location'),
    name: z.string(),
    locLat: z.string(),
    locLong: z.string(),
    locTrigger: z.enum(['on_enter', 'on_leave']),
    radius: z.number().int(),
}).passthrough();
export const AbsoluteReminderSchema = ReminderBaseSchema.extend({
    type: z.literal('absolute'),
    due: DueDateSchema,
}).passthrough();
export const RelativeReminderSchema = ReminderBaseSchema.extend({
    type: z.literal('relative'),
    minuteOffset: z.number().int(),
    due: DueDateSchema.optional(),
}).passthrough();
export const ReminderSchema = z.discriminatedUnion('type', [
    LocationReminderSchema,
    AbsoluteReminderSchema,
    RelativeReminderSchema,
]);
