// @doist/todoist-api-typescript/dist/esm/types/sync/resources/suggestions.js
import { z } from 'zod';
export const TemplateSuggestionSchema = z
    .object({
    id: z.string(),
    name: z.string(),
    templateType: z.enum(['project', 'setup']),
})
    .passthrough();
export const WorkspaceTemplateSuggestionSchema = TemplateSuggestionSchema.extend({
    workspaceId: z.string().nullable(),
});
const SyncTemplateSuggestionsSchema = z
    .object({
    type: z.enum(['templates', 'most_used_user_templates']),
    content: z.object({
        templates: z.array(TemplateSuggestionSchema),
        locale: z.string(),
    }),
    isDeleted: z.boolean(),
})
    .passthrough();
const SyncWorkspaceTemplateSuggestionsSchema = z
    .object({
    type: z.literal('most_used_workspace_templates'),
    content: z.object({
        templates: z.array(WorkspaceTemplateSuggestionSchema),
        locale: z.string(),
    }),
    isDeleted: z.boolean(),
})
    .passthrough();
export const SuggestionSchema = z.union([
    SyncWorkspaceTemplateSuggestionsSchema,
    SyncTemplateSuggestionsSchema,
]);
