// @doist/todoist-api-typescript/dist/cjs/consts/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENDPOINT_WORKSPACE_USERS = exports.ENDPOINT_WORKSPACE_PLAN_DETAILS = exports.ENDPOINT_WORKSPACE_LOGO = exports.ENDPOINT_WORKSPACE_JOIN = exports.ENDPOINT_WORKSPACE_INVITATIONS_DELETE = exports.ENDPOINT_WORKSPACE_INVITATIONS_ALL = exports.ENDPOINT_WORKSPACE_INVITATIONS = exports.ENDPOINT_REVOKE = exports.ENDPOINT_GET_TOKEN = exports.ENDPOINT_AUTHORIZATION = exports.ENDPOINT_SYNC = exports.ENDPOINT_SYNC_QUICK_ADD = exports.ENDPOINT_REST_PROJECTS_MOVE_TO_PERSONAL = exports.ENDPOINT_REST_PROJECTS_MOVE_TO_WORKSPACE = exports.PROJECT_UNARCHIVE = exports.PROJECT_ARCHIVE = exports.ENDPOINT_REST_UPLOADS = exports.ENDPOINT_REST_ACTIVITIES = exports.ENDPOINT_REST_PRODUCTIVITY = exports.ENDPOINT_REST_USER = exports.ENDPOINT_REST_PROJECT_COLLABORATORS = exports.ENDPOINT_REST_PROJECTS_ARCHIVED = exports.ENDPOINT_REST_PROJECTS_SEARCH = exports.ENDPOINT_REST_PROJECTS = exports.ENDPOINT_REST_TASK_MOVE = exports.ENDPOINT_REST_TASK_REOPEN = exports.ENDPOINT_REST_TASK_CLOSE = exports.ENDPOINT_REST_COMMENTS = exports.ENDPOINT_REST_LABELS_SHARED_REMOVE = exports.ENDPOINT_REST_LABELS_SHARED_RENAME = exports.ENDPOINT_REST_LABELS_SHARED = exports.ENDPOINT_REST_LABELS_SEARCH = exports.ENDPOINT_REST_LABELS = exports.ENDPOINT_REST_SECTIONS_SEARCH = exports.ENDPOINT_REST_SECTIONS = exports.ENDPOINT_REST_TASKS_COMPLETED_SEARCH = exports.ENDPOINT_REST_TASKS_COMPLETED_BY_DUE_DATE = exports.ENDPOINT_REST_TASKS_COMPLETED_BY_COMPLETION_DATE = exports.ENDPOINT_REST_TASKS_FILTER = exports.ENDPOINT_REST_TASKS = exports.API_BASE_URI = exports.API_VERSION = exports.TODOIST_WEB_URI = void 0;
exports.getSyncBaseUri = getSyncBaseUri;
exports.getAuthBaseUri = getAuthBaseUri;
exports.getWorkspaceInvitationAcceptEndpoint = getWorkspaceInvitationAcceptEndpoint;
exports.getWorkspaceInvitationRejectEndpoint = getWorkspaceInvitationRejectEndpoint;
exports.getWorkspaceActiveProjectsEndpoint = getWorkspaceActiveProjectsEndpoint;
exports.getWorkspaceArchivedProjectsEndpoint = getWorkspaceArchivedProjectsEndpoint;
const BASE_URI = 'https://api.todoist.com';
const TODOIST_URI = 'https://todoist.com';
exports.TODOIST_WEB_URI = 'https://app.todoist.com/app';
// The API version is not configurable, to ensure
// compatibility between the API and the client.
exports.API_VERSION = 'v1';
exports.API_BASE_URI = `/api/${exports.API_VERSION}/`;
const API_AUTHORIZATION_BASE_URI = '/oauth/';
function getSyncBaseUri(domainBase = BASE_URI) {
    return new URL(exports.API_BASE_URI, domainBase).toString();
}
function getAuthBaseUri(domainBase = TODOIST_URI) {
    return new URL(API_AUTHORIZATION_BASE_URI, domainBase).toString();
}
exports.ENDPOINT_REST_TASKS = 'tasks';
exports.ENDPOINT_REST_TASKS_FILTER = exports.ENDPOINT_REST_TASKS + '/filter';
exports.ENDPOINT_REST_TASKS_COMPLETED_BY_COMPLETION_DATE = exports.ENDPOINT_REST_TASKS + '/completed/by_completion_date';
exports.ENDPOINT_REST_TASKS_COMPLETED_BY_DUE_DATE = exports.ENDPOINT_REST_TASKS + '/completed/by_due_date';
exports.ENDPOINT_REST_TASKS_COMPLETED_SEARCH = 'completed/search';
exports.ENDPOINT_REST_SECTIONS = 'sections';
exports.ENDPOINT_REST_SECTIONS_SEARCH = exports.ENDPOINT_REST_SECTIONS + '/search';
exports.ENDPOINT_REST_LABELS = 'labels';
exports.ENDPOINT_REST_LABELS_SEARCH = exports.ENDPOINT_REST_LABELS + '/search';
exports.ENDPOINT_REST_LABELS_SHARED = exports.ENDPOINT_REST_LABELS + '/shared';
exports.ENDPOINT_REST_LABELS_SHARED_RENAME = exports.ENDPOINT_REST_LABELS_SHARED + '/rename';
exports.ENDPOINT_REST_LABELS_SHARED_REMOVE = exports.ENDPOINT_REST_LABELS_SHARED + '/remove';
exports.ENDPOINT_REST_COMMENTS = 'comments';
exports.ENDPOINT_REST_TASK_CLOSE = 'close';
exports.ENDPOINT_REST_TASK_REOPEN = 'reopen';
exports.ENDPOINT_REST_TASK_MOVE = 'move';
exports.ENDPOINT_REST_PROJECTS = 'projects';
exports.ENDPOINT_REST_PROJECTS_SEARCH = exports.ENDPOINT_REST_PROJECTS + '/search';
exports.ENDPOINT_REST_PROJECTS_ARCHIVED = exports.ENDPOINT_REST_PROJECTS + '/archived';
exports.ENDPOINT_REST_PROJECT_COLLABORATORS = 'collaborators';
exports.ENDPOINT_REST_USER = 'user';
exports.ENDPOINT_REST_PRODUCTIVITY = exports.ENDPOINT_REST_TASKS + '/completed/stats';
exports.ENDPOINT_REST_ACTIVITIES = 'activities';
exports.ENDPOINT_REST_UPLOADS = 'uploads';
exports.PROJECT_ARCHIVE = 'archive';
exports.PROJECT_UNARCHIVE = 'unarchive';
exports.ENDPOINT_REST_PROJECTS_MOVE_TO_WORKSPACE = exports.ENDPOINT_REST_PROJECTS + '/move_to_workspace';
exports.ENDPOINT_REST_PROJECTS_MOVE_TO_PERSONAL = exports.ENDPOINT_REST_PROJECTS + '/move_to_personal';
exports.ENDPOINT_SYNC_QUICK_ADD = exports.ENDPOINT_REST_TASKS + '/quick';
exports.ENDPOINT_SYNC = 'sync';
exports.ENDPOINT_AUTHORIZATION = 'authorize';
exports.ENDPOINT_GET_TOKEN = 'access_token';
exports.ENDPOINT_REVOKE = 'revoke';
// Workspace endpoints
exports.ENDPOINT_WORKSPACE_INVITATIONS = 'workspaces/invitations';
exports.ENDPOINT_WORKSPACE_INVITATIONS_ALL = 'workspaces/invitations/all';
exports.ENDPOINT_WORKSPACE_INVITATIONS_DELETE = 'workspaces/invitations/delete';
exports.ENDPOINT_WORKSPACE_JOIN = 'workspaces/join';
exports.ENDPOINT_WORKSPACE_LOGO = 'workspaces/logo';
exports.ENDPOINT_WORKSPACE_PLAN_DETAILS = 'workspaces/plan_details';
exports.ENDPOINT_WORKSPACE_USERS = 'workspaces/users';
// Workspace invitation actions (require invite_code parameter)
function getWorkspaceInvitationAcceptEndpoint(inviteCode) {
    return `workspaces/invitations/${inviteCode}/accept`;
}
function getWorkspaceInvitationRejectEndpoint(inviteCode) {
    return `workspaces/invitations/${inviteCode}/reject`;
}
// Workspace projects (require workspace_id parameter)
function getWorkspaceActiveProjectsEndpoint(workspaceId) {
    return `workspaces/${workspaceId}/projects/active`;
}
function getWorkspaceArchivedProjectsEndpoint(workspaceId) {
    return `workspaces/${workspaceId}/projects/archived`;
}
